### DISCUSSION TITLE
Failing to set width/height properties on canvas when it is inside a grid

### DISCUSSION THREAD

**Discussion by Basnederveen** [AvaloniaUI/Avalonia#11184] [Q&A]
Hi all, I'm new to avalonia and have been breaking my head over this problem for 2 days now and I do not know what to do anymore. I think I am just missing something small.

I have a custom canvas page using the PanAndZoom control. Here I want to draw some lines from a viewmodel. The width/height of the canvas is coming from the CanvasViewModel 

`
<Canvas Background="LightGray"
	Width="{Binding CanvasPage.Width}"
	Height="{Binding CanvasPage.Height}">
`

Everything is working fine when I place the control inside a tabcontrol:

`	
<Panel>
  <TabControl x:Name="TabControl" Classes="sidebar" Margin="0 40 0 20">
	  <TabItem Header="Create grid">
		  <pages:CanvasPage Margin="0 0 50 0"/>
	  </TabItem>
  </TabControl>
</Panel>
`

The desiredsize of the canvaspage shows as the calculated values (bounds of the contents + offset). This is a large value, that's why i'm using the zoomcontrol. But I want to keep 1 px = 1mm.

![image](https://user-images.githubusercontent.com/16050771/235297310-e942c80a-7f30-43ee-80d7-a501464eef2e.png)

When I place the control inside a grid, the size is not correct anymore.

`
<Panel Grid.Row="1" Grid.Column="1">
	<pages:CanvasPage ></pages:CanvasPage>
</Panel>
`

![image](https://user-images.githubusercontent.com/16050771/235297424-3cfcb28e-3005-4cf8-a489-a5f331668396.png)

Does anyone have an idea what could be going wrong? Can supply more code if required, only the container of the canvaspage seems to be the problem.

Below the complete CanvasPage code:

<UserControl xmlns="https://github.com/avaloniaui"
             xmlns:x="http://schemas.microsoft.com/winfx/2006/xaml"
             xmlns:d="http://schemas.microsoft.com/expression/blend/2008"
             xmlns:mc="http://schemas.openxmlformats.org/markup-compatibility/2006"
             xmlns:paz="using:Avalonia.Controls.PanAndZoom"
             xmlns:vm="using:Olsthoorn.DesktopUI.ViewModels"
			 mc:Ignorable="d" d:DesignWidth="800" d:DesignHeight="450"
             x:Class="Olsthoorn.DesktopUI.Views.Pages.CanvasPage">

	<UserControl.Resources>
		<VisualBrush x:Key="TemplateCheckerBoard" TileMode="Tile" SourceRect="0,0,20,20" DestinationRect="0,0,20,20">
			<VisualBrush.Visual>
				<Path Data="M 0,0 L 0,10 L 10,10 L 10,0 Z M 10,10 L 10,20 L 20,20 L 20,10 Z" Fill="#FFF0F0F0" />
			</VisualBrush.Visual>
		</VisualBrush>
	</UserControl.Resources>

	<Panel Background="Transparent">
		<Panel Background="#FFF5F5F5" />
		<Panel Background="{DynamicResource TemplateCheckerBoard}" />
		<Panel ClipToBounds="True">
			<paz:ZoomBorder Name="ZoomBorder"
							Stretch="None"
							ZoomSpeed="1.5"
							PanButton="Middle"
							EnablePan="True"
							EnableZoom="True"
							Background="Transparent"
							ClipToBounds="True"
							VerticalAlignment="Stretch"
							HorizontalAlignment="Stretch">

				<ItemsControl Items="{Binding CanvasPage.Lines}">

					<ItemsControl.ItemsPanel>
						<ItemsPanelTemplate>
							<Canvas Background="LightGray"
									Width="{Binding CanvasPage.Width}"
									Height="{Binding CanvasPage.Height}">
								<Canvas.RenderTransform>
									<TransformGroup>
										<MatrixTransform Matrix="1 0 0 -1 0 0" />
									</TransformGroup>
								</Canvas.RenderTransform>
							</Canvas>
						</ItemsPanelTemplate>
					</ItemsControl.ItemsPanel>

					<ItemsControl.DataTemplates>
						<DataTemplate DataType="vm:LineViewModel">

							<Line StartPoint="{Binding Start}"
								  EndPoint="{Binding End}"
								  Stroke="Black" StrokeThickness="20">

							</Line>
						</DataTemplate>
					</ItemsControl.DataTemplates>
				</ItemsControl>
			</paz:ZoomBorder>
		</Panel>
	</Panel>
</UserControl>

EDIT: Here is a a repo where it can be tried. Edit mainview.xaml code to see what happens when the canvaspage is in the tabcontrol or in the grid. 
https://github.com/Basnederveen/CanvasHeightWidthProblem.git

**Comment by timunie** [comment]
@Basnederveen I think this is a probably a question for the PanAndZoom-control repo. Can you file this issue there please?

**Reply by Basnederveen** [reply]
Thanks for the reply. Sure will do.
